In [66]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
# ML Models
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LinearRegression , Ridge
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.ensemble import RandomForestClassifier
# Ignore warnings
import warnings
warnings.filterwarnings("ignore")

In [67]:
# Read data set into pandas data frame.
df = pd.read_csv("Cancer.csv")

In [68]:
# Display first 5 rows of the data
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [69]:
# Check for nulls and data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

In [70]:
# Remove unnamed column as it doesn't give value.
df.drop(['Unnamed: 32'], axis = 1, inplace = True) 

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

In [72]:
# Check Duplicates
df.duplicated().sum()

0

In [73]:
# Check Correlation
df.corr()

id  radius_mean  texture_mean  perimeter_mean  \
id                       1.000000     0.074626      0.099770        0.073159   
radius_mean              0.074626     1.000000      0.323782        0.997855   
texture_mean             0.099770     0.323782      1.000000        0.329533   
perimeter_mean           0.073159     0.997855      0.329533        1.000000   
area_mean                0.096893     0.987357      0.321086        0.986507   
smoothness_mean         -0.012968     0.170581     -0.023389        0.207278   
compactness_mean         0.000096     0.506124      0.236702        0.556936   
concavity_mean           0.050080     0.676764      0.302418        0.716136   
concave points_mean      0.044158     0.822529      0.293464        0.850977   
symmetry_mean           -0.022114     0.147741      0.071401        0.183027   
fractal_dimension_mean  -0.052511    -0.311631     -0.076437       -0.261477   
radius_se                0.143048     0.679090      0.275869        0.691765   
texture_se              -0.007526    -0.097317      0.386358       -0.086761   
perimeter_se             0.137331     0.674172      0.281673        0.693135   
area_se                  0.177742     0.735864      0.259845        0.744983   
smoothness_se            0.096781    -0.222600      0.006614       -0.202694   
compactness_se           0.033961     0.206000      0.191975        0.250744   
concavity_se             0.055239     0.194204      0.143293        0.228082   
concave points_se        0.078768     0.376169      0.163851        0.407217   
symmetry_se             -0.017306    -0.104321      0.009127       -0.081629   
fractal_dimension_se     0.025725    -0.042641      0.054458       -0.005523   
radius_worst             0.082405     0.969539      0.352573        0.969476   
texture_worst            0.064720     0.297008      0.912045        0.303038   
perimeter_worst          0.079986     0.965137      0.358040        0.970387   
area_worst               0.107187     0.941082      0.343546        0.941550   
smoothness_worst         0.010338     0.119616      0.077503        0.150549   
compactness_worst       -0.002968     0.413463      0.277830        0.455774   
concavity_worst          0.023203     0.526911      0.301025        0.563879   
concave points_worst     0.035174     0.744214      0.295316        0.771241   
symmetry_worst          -0.044224     0.163953      0.105008        0.189115   
fractal_dimension_worst -0.029866     0.007066      0.119205        0.051019   

                         area_mean  smoothness_mean  compactness_mean  \
id                        0.096893        -0.012968          0.000096   
radius_mean               0.987357         0.170581          0.506124   
texture_mean              0.321086        -0.023389          0.236702   
perimeter_mean            0.986507         0.207278          0.556936   
area_mean                 1.000000         0.177028          0.498502   
smoothness_mean           0.177028         1.000000          0.659123   
compactness_mean          0.498502         0.659123          1.000000   
concavity_mean            0.685983         0.521984          0.883121   
concave points_mean       0.823269         0.553695          0.831135   
symmetry_mean             0.151293         0.557775          0.602641   
fractal_dimension_mean   -0.283110         0.584792          0.565369   
radius_se                 0.732562         0.301467          0.497473   
texture_se               -0.066280         0.068406          0.046205   
perimeter_se              0.726628         0.296092          0.548905   
area_se                   0.800086         0.246552          0.455653   
smoothness_se            -0.166777         0.332375          0.135299   
compactness_se            0.212583         0.318943          0.738722   
concavity_se              0.207660         0.248396          0.570517   
concave points_se         0.372320         0.380676          0.642262   
symmetry_s

In [74]:
# Drop outliers
mean = df.mean()
std_dev = df.std()
maxx = mean + 3 * std_dev
minn = mean - 3 * std_dev
outliers = df[(df > maxx) | (df < minn)].stack()
df = df.drop(outliers.index.get_level_values(0))

In [75]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 487 entries, 1 to 566
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       487 non-null    int64  
 1   diagnosis                487 non-null    object 
 2   radius_mean              487 non-null    float64
 3   texture_mean             487 non-null    float64
 4   perimeter_mean           487 non-null    float64
 5   area_mean                487 non-null    float64
 6   smoothness_mean          487 non-null    float64
 7   compactness_mean         487 non-null    float64
 8   concavity_mean           487 non-null    float64
 9   concave points_mean      487 non-null    float64
 10  symmetry_mean            487 non-null    float64
 11  fractal_dimension_mean   487 non-null    float64
 12  radius_se                487 non-null    float64
 13  texture_se               487 non-null    float64
 14  perimeter_se             4

In [76]:
# Split the data into labels and features
X = df.drop(columns='diagnosis')
y = df[['diagnosis']]
y.head()

diagnosis
1         M
2         M
4         M
5         M
6         M

In [77]:
# Normalize data Using Z-score for better results
for column in X:
    X[column] = (X[column] -X[column].mean()) / X[column].std()    

    
# view normalized data   
display(X)

id  radius_mean  texture_mean  perimeter_mean  area_mean  \
1   -0.422611     2.222574     -0.275188        2.084027   2.478382   
2    2.571686     1.931377      0.599998        1.943677   2.046082   
4    2.573749     2.129920     -1.137800        2.190499   2.376458   
5   -0.422565    -0.464384     -0.795773       -0.351776  -0.505192   
6   -0.422545     1.454872      0.280605        1.440352   1.473197   
..        ...          ...           ...             ...        ...   
560 -0.419641     0.065066      2.083791        0.074598  -0.071837   
563 -0.419611     2.338391      1.565721        2.572833   2.552190   
564 -0.419600     2.550171      0.886697        2.524436   3.016122   
565 -0.419591     2.076975      2.360430        2.001753   2.249931   
566 -0.419581     0.908877      2.317677        0.893470   0.833884   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
1          -0.810290         -0.397167        0.205598             0.829361   
2           1.129111          1.578582        1.995202             2.608723   
4           0.403591          0.919675        2.004920             1.881320   
5           2.548946          1.824153        1.353860             1.159774   
6          -0.038742          0.341004        0.623442             0.947410   
..               ...               ...             ...                  ...   
560         0.324798          0.428534       -0.479148            -0.006844   
563         1.152515          3.127379        3.938664             3.209754   
564         1.238329          0.508770        2.748294             2.947766   
565         0.208559          0.204846        1.130362             1.684367   
566        -0.825112          0.178100        0.296455             0.300761   

     symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
1         0.151269  ...      2.232985      -0.302549         1.954625   
2         1.267057  ...      1.887173       0.070503         1.730729   
4         0.138244  ...      1.636337      -1.488573         1.720067   
5         1.345206  ...     -0.085417      -0.242720        -0.014240   
6         0.073121  ...      1.719137       0.445314         1.755606   
..             ...  ...           ...            ...              ...   
560      -1.042668  ...     -0.126817       1.414898        -0.127966   
563       1.614384  ...      2.062514       0.753258         2.676069   
564      -0.222107  ...      2.345009       0.223595         2.214061   
565      -0.109226  ...      1.916396       2.308815         1.819577   
566      -0.812563  ...      0.769372       1.582067         0.813820   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
1      2.533921         -0.334161          -0.360026         0.000952   
2      1.984206          0.641266           1.527909         1.219008   
4      1.685980          0.309811          -0.214006         0.924994   
5     -0.168810          2.284340           2.324667         1.715447   
6      1.754972          0.631796           0.203418         0.798988   
..          ...               ...                ...              ...   
560   -0.246482         -0.319956          -0.044180        -0.634911   
563    2.229018          0.466068           1.481088         2.441148   
564    2.691937          0.480274          -0.164011         0.987414   
565    2.033168         -0.675087          -0.315585         0.467056   
566    0.682248         -0.802934           0.614495         0.576728   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
1                1.369769       -0.181461                 0.513141  
2                2.352410        1.552452                 0.413144  
4                0.964645       -0.957000                -0.336837  
5                1.164621        2.299863                 2.970021  
6                1.493892        0.447409                 0.142317  
..                    ...             ...

In [78]:
# Change string values to binary so the classifier could understand them
y['diagnosis'] = y['diagnosis'].replace(['M','B'],['1','0'])

In [79]:
# Splitting the data into training and testing
x_train,x_test,y_train,y_test= train_test_split(X,y,test_size=0.2,random_state= 42)

In [80]:
# Linear Regression Model
cl1 = LinearRegression().fit(x_train,y_train)
cl1_pred = cl1.predict(x_test)

In [81]:
# Ridge Model
cl2 = Ridge(alpha = 1).fit(x_train,y_train)
cl2_pred = cl2.predict(x_test)

In [82]:
# Grid Search Model  using decision tree regressor
params = [{'max_depth':list(range(5,20)),'min_samples_split':list(range(2,15)),'min_samples_leaf':[2,3,4]}]
cl3 = GridSearchCV(estimator = DecisionTreeRegressor(), param_grid=params, cv=10, n_jobs=-1).fit(x_train,y_train)
cl3_pred = cl3.predict(x_test)

In [83]:
# Decision Tree Model
cl4 = DecisionTreeRegressor().fit(x_train,y_train)
cl4_pred = cl4.predict(x_test)

In [84]:
# Random Forest Model
cl5 = RandomForestClassifier().fit(x_train,y_train)
cl5_pred = cl5.predict(x_test)

In [85]:
print("Linear regression r2 score: ",r2_score(y_test,cl1_pred))
print("Linear regression MSE: ",mean_squared_error(y_test,cl1_pred))
print("--------------------------")
print("Ridge r2 score: ",r2_score(y_test,cl2_pred))
print("Ridge MSE: ",mean_squared_error(y_test,cl2_pred))
print("--------------------------")
print("Grid Search r2 score: ",r2_score(y_test,cl3_pred))
print("Grid Search MSE: ",mean_squared_error(y_test,cl3_pred))
print("--------------------------")
print("Decision Tree r2 score: ",r2_score(y_test,cl4_pred))
print("Decision Tree MSE: ",mean_squared_error(y_test,cl4_pred))
print("--------------------------")
print("Random Forest r2 score: ",r2_score(y_test,cl5_pred))
print("Random Forest MSE: ",mean_squared_error(y_test,cl5_pred))

Linear regression r2 score:  0.7155135044341661
Linear regression MSE:  0.06815945713213076
--------------------------
Ridge r2 score:  0.7179051100697802
Ridge MSE:  0.06758645790602205
--------------------------
Grid Search r2 score:  0.6250171423052779
Grid Search MSE:  0.08984126984126983
--------------------------
Decision Tree r2 score:  0.5315080399826162
Decision Tree MSE:  0.11224489795918367
--------------------------
Random Forest r2 score:  0.8722294654498044
Random Forest MSE:  0.030612244897959183


### The best model as we can see is the random forest model with 87% r2 score and a 3% error.